dotfiles/emacs/profiles/main/config.org
#+TITLE: Ceci n'est pas une pipe

* Documentation
** Cheatsheet
*** Core
|--------------------------------+-----------------|
| Action                         | Key             |
|--------------------------------+-----------------|
| Help                           | <C-?> or <C-M-? |
| Command panel (CSP             | <C-S-p>         |
| Switch to project              | <leader>p       |
| Navigate to file               | <C-p>           |
| Navigate to file (text search) | <C-f>           |
| Navigate to buffer             | ;               |
| Navigate to previous buffer    | DEL             |
| Navigate to line in buffer     | <leader>l       |
| Save buffer                    | <C-s>           |
| Kill buffer                    | <C-w>           |
| Edit configuration             | <leader>ee      |
| Reload configuration           | <C-S-r>         |
| Profiler                       | <leader>P       |
| Explore top-level mappings     | <leader>?       |
| Open / focus file manager      | -               |
| Quit                           | <leader>q       |
|--------------------------------+-----------------|
*** Text editing
|---------------------+-----------------|
| Action              | Key             |
|---------------------+-----------------|
| Move line up / down | <M-k> and <M-j> |
|---------------------+-----------------|
*** Development
|---------------------------+------------|
| Action                    | Key        |
|---------------------------+------------|
| Go to definition          | gd         |
| Go to next reference      | gr         |
| Go to previous reference  | gR         |
| Glance documentation      | K          |
| Go to next diagnostic     | gp         |
| Go to previous diagnostic | gP         |
| Go to symbol in buffer    | <leader>r  |
| Execute code action       | <leader>ac |
| Rename symbol             | <leader>er |
|---------------------------+------------|
*** Org
|-----------------------------------+----------|
| Action                            | Key      |
|-----------------------------------+----------|
| Insert same-level item below      | <C-CR>   |
| Insert same-level todo item below | <C-S-CR> |
| Cycle item status                 | <C-SPC>  |
|-----------------------------------+----------|
*** Wiki
|---------------------+---------|
| Action              | Key     |
|---------------------+---------|
| Find node (default) | <M-n>   |
| Find project node   | <M-p>   |
| Insert link to node | <C-c>ni |
| Capture             | <C-c>c  |
| Refile              | <C-c>r  |
| Archive             | <C-c>z  |
|---------------------+---------|
*** Agenda
|-----------------------------------------+------------------------+-----------------+---------|
| Action                                  | Key                    | Command         | Context |
|-----------------------------------------+------------------------+-----------------+---------|
| Agenda prefix                           | <C-c>a or <leader>a    | org-agenda      | Global  |
| Open current agenda                     | <C-c>aa or <leader>aa  | org-agenda-list | Global  |
| Open current agenda and todos           | <C-c>an  or <leader>an |                 | Global  |
| Open todo list                          | <C-c>at  or <leader>at | org-todo-list   | Global  |
| Schedule item (pick date: shift+arrows) | <C-c>s                 | org-schedule    | Org     |
| Clock in                                | <C-c>i                 | org-clock-in    | Org     |
| Clock out                               | <C-c>o                 | org-clock-out   | Org     |
| Show clock stats                        | <C-c><C-x><C-c>        | org-columns     | Org     |
|-----------------------------------------+------------------------+-----------------+---------|
*** Journal
|---------------+------------|
| Action        | Key        |
|---------------+------------|
| Go to today   | <leader>dd |
| Go to date    | <leader>dc |
| Capture today | <leader>da |
|---------------+------------|
** Packages
*** Package management
Packages are managed by ~straight.el~.
Commands:
~straight-pull-all~ - update packags
~straight-freeze-versions~ - lock package versions
~straight-thaw-versions~ - restore to locked versions
*** Hooks
**** ~(with-eval-after-load 'library ...)~
Evaluates body at the end of loading the ~library~, *each time* the library is loaded.
If the library is already loaded, evaluates body right away.
*** ~general.el~
#+BEGIN_SRC emacs-lisp :tangle no
;; Define mapping
(general-def
  "M-n" 'org-next-visible-heading)
(general-def org-mode-map
  "M-n" 'org-next-visible-heading)
(general-def 'normal org-mode-map
  "M-n" 'org-next-visible-heading)
(general-def
  :states '(normal motion emacs)
  "M-n" 'org-next-visible-heading)

;; Define for mode
(general-define-key
 :states '(normal visual)
 :keymaps 'org-mode-map
 "M-n" 'org-next-visible-heading
 "M-p" 'org-previous-visible-heading)
(general-evil-define-key
    :states '(normal visual)
  :keymaps 'org-mode-map
  "M-n" 'org-next-visible-heading
  "M-p" 'org-previous-visible-heading)

;; Define for keymap
(general-emacs-define-key org-mode-map
  "M-n" 'org-next-visible-heading
  "M-p" 'org-previous-visible-heading)
(general-define-key
 :keymaps 'org-mode-map
 "M-n" 'org-next-visible-heading
 "M-p" 'org-previous-visible-heading)
(with-eval-after-load 'org-mode
  (define-key org-mode-map (kbd "M-n") 'org-next-visible-heading)
  )

;; Unbind
(general-unbind 'insert
  "C-v"
  "C-k"
  "C-y"
  "C-e")
#+END_SRC
*** ~use-package~
- ~:init~ - before loading
- ~:config~ - after loading
- ~:custom~ - set customizable variables
- ~:bind~  - create binds after the module is loaded
  : ("C-x c" . company-mode)
  : :map isearch-mode-map
  : ("C-x c" . company-mode)
- ~:magic ("%PDF" . pdf-view-mode)~ - run on file header match
- ~:after (...)~ - set loading order
  : (:all foo bar) or just (foo bar)
  : (:any foo bar)
  : (:all (:any foo bar) (:any baz quux))

* Core
** No littering :pkg:
#+BEGIN_SRC emacs-lisp
(setq custom-file (make-temp-file ""))
(setq
 user-emacs-directory (expand-file-name "~/.cache/emacs/")
 url-history-file (expand-file-name "url/history" user-emacs-directory))

(use-package no-littering
  :straight t)
#+END_SRC
** Startup profiling
#+BEGIN_SRC emacs-lisp
(add-hook
 'emacs-startup-hook
 (lambda ()
   (message
    "*** Emacs loaded in %s with %d garbage collections."
    (format "%.2f seconds" (float-time
                            (time-subtract after-init-time before-init-time)))
    gcs-done)))
#+END_SRC
** Keyboard
*** Setup evil and undo
#+BEGIN_SRC emacs-lisp
;; (use-package undo-fu)
(use-package evil
  :straight t
  :init
  ;; (setq evil-undo-system 'undo-fu)
  (setq evil-want-integration t)
  (setq evil-want-keybinding nil)
  (setq evil-want-C-u-scroll t)
  (setq evil-want-C-i-jump t)
  (setq evil-want-C-u-delete t)
  (setq evil-want-C-w-delete t)
  (setq evil-want-Y-yank-to-eol t)
  (setq evil-respect-visual-line-mode t)
  (setq evil-undo-system 'undo-tree)
  (setq evil-search-module 'evil-search)
  (setq evil-split-window-below t)
  (setq evil-vsplit-window-right t)
  :config
  (evil-mode 1)
  (evil-global-set-key 'motion "j" 'evil-next-visual-line)
  (evil-global-set-key 'motion "k" 'evil-previous-visual-line)
  (evil-set-initial-state 'messages-buffer-mode 'normal)
  (evil-set-initial-state 'dashboard-mode 'normal))
#+END_SRC
*** Setup general.el                                                    :pkg:
#+BEGIN_SRC emacs-lisp
(use-package general
  :straight t
  :config
  (general-evil-setup t))
#+END_SRC
*** Setup leader and provide ~lib/mapleader~
#+BEGIN_SRC emacs-lisp
(general-create-definer lib/mapleader
  :keymaps 'override
  :prefix "SPC"
  :states '(normal motion emacs))
#+END_SRC
*** Reset bindings
https://github.com/emacs-evil/evil/blob/master/evil-maps.el
#+BEGIN_SRC emacs-lisp
;; emacs
(global-set-key (kbd "C-SPC") nil)

;; evil
(with-eval-after-load 'evil-maps
  ;; remove default mappings
  (define-key evil-normal-state-map (kbd "DEL") nil)
  (define-key evil-normal-state-map (kbd "<C-return>") nil)
  (define-key evil-normal-state-map (kbd "\C-p") nil)
  (define-key evil-normal-state-map (kbd "\C-w") nil)
  (define-key evil-insert-state-map (kbd "\C-w") nil)
  (define-key evil-motion-state-map (kbd "RET") nil)
  (define-key evil-motion-state-map (kbd "\C-f") nil)
  (define-key evil-motion-state-map (kbd "-") nil)
  (define-key evil-motion-state-map (kbd ";") nil)
  (define-key evil-motion-state-map (kbd "\C-w") nil)
  (define-key evil-emacs-state-map (kbd "\C-w") nil)
  ;; fix tab
  (define-key evil-insert-state-map (kbd "TAB") 'tab-to-tab-stop)
  )
#+END_SRC
*** Remap help (~<C-?>~, ~<C-M-?>~)
#+BEGIN_SRC emacs-lisp
(global-set-key (kbd "C-?") 'help-command)
(global-set-key (kbd "C-M-?") 'help-command)
#+END_SRC
*** Remap universal argument
#+BEGIN_SRC emacs-lisp
(general-def
  :keymaps 'universal-argument-map
  "M-u" 'universal-argument-more)
(general-def
  :keymaps 'override
  :states '(normal motion emacs insert visual)
  "M-u" 'universal-argument)
#+END_SRC
** Backups
#+BEGIN_SRC emacs-lisp
(setq make-backup-files nil)
(setq auto-save-default nil)
(setq create-lockfiles nil)
(setq backup-inhibited t)
#+END_SRC
** Scrolling
#+BEGIN_SRC  emacs-lisp
(setq scroll-conservatively 101)
(setq scroll-step 1)
(setq scroll-preserve-screen-position t)
(setq scroll-error-top-bottom t)

(setq mouse-wheel-scroll-amount '(1 ((shift) . 1)))
(setq mouse-wheel-progressive-speed nil)
(setq mouse-wheel-follow-mouse 't)
#+END_SRC
** Code style
#+BEGIN_SRC emacs-lisp
(setq tab-always-indent nil)
(setq-default default-tab-width 2)
(setq-default tab-width 2)
(setq-default indent-tabs-mode nil)

(setq-default evil-shift-width tab-width)
(setq-default evil-indent-convert-tabs nil)
(setq-default evil-shift-round nil)
#+END_SRC
** Misc settings
#+BEGIN_SRC emacs-lisp
(setq warning-minimum-level :error)
(setq enable-recursive-minibuffers t)
(normal-erase-is-backspace-mode 1)
(setq visible-bell nil)
(setq use-dialog-box nil)
(setq large-file-warning-threshold nil)
(setq vc-follow-symlinks t)
(setq ad-redefinition-action 'accept)
(global-auto-revert-mode 1)
(setq global-auto-revert-non-file-buffers t)
(defalias 'yes-or-no-p 'y-or-n-p)
(setq save-interprogram-paste-before-kill t)
(setq confirm-nonexistent-file-or-buffer nil)
(setq revert-without-query '(".*"))
(setq vc-follow-symlinks t)
(setq-default bidi-display-reordering nil)
(remove-hook 'find-file-hooks 'vc-find-file-hook)
(setq async-shell-command-buffer 'new-buffer)
(setq shell-command-switch "-ic")
(global-unset-key [(control z)])
(global-unset-key [(control x)(control z)])
(set-frame-parameter (selected-frame) 'fullscreen 'maximized)
(add-to-list 'default-frame-alist '(fullscreen . maximized))
(setq x-underline-at-descent-line t)
#+END_SRC
** Utils
*** ~lib/time "time begin|end"~
#+BEGIN_SRC emacs-lisp
(defun lib/log (string)
  "Print out STRING and calculate length of init."
  (message string)
  (if (not (string= "end" (substring string -3)))
      (setq my/init-audit-message-begin (current-time))
    (message
     "%s seconds"
     (time-to-seconds
      (time-subtract
       (current-time)
       my/init-audit-message-begin))))
  nil)
#+END_SRC
** Init
#+BEGIN_SRC emacs-lisp
(cd "~/brain/wiki")
#+END_SRC
* Packages
*** alert
#+BEGIN_SRC emacs-lisp
(use-package alert
  :straight t
  :commands alert
  :config
  (setq alert-default-style 'notifications))
#+END_SRC
*** all-the-icons
#+BEGIN_SRC emacs-lisp
(use-package all-the-icons
  :straight t
  :if (display-graphic-p))
(use-package nerd-icons)
#+END_SRC
*** company
#+BEGIN_SRC emacs-lisp
(use-package company
  :straight t
  :init
  (global-company-mode)

  :config
  (setq company-global-modes '(not eshell-mode gud-mode))
  (setq company-minimum-prefix-length 0)
  (setq company-idle-delay 0.0)
  (setq company-require-match nil)
  (setq company-dabbrev-ignore-case nil)
  (setq company-dabbrev-downcase nil)
  (setq company-selection-wrap-around t)
  (setq company-tooltip-align-annotations t)
  (setq company-tooltip-flip-when-above t)
  (setq company-tooltip-limit 20)
  (company-tng-configure-default))

(use-package company-flx
  :straight t
  :after company
  :init
  (company-flx-mode)

  :config
  (setq company-flx-limit 100))

(use-package company-lsp
  :straight t
  :requires company
  ;; :config
  ;; (push 'company-lsp company-backends)
  ;; (setq company-transformers nil
  ;;       company-lsp-async t
  ;;       company-lsp-cache-candidates nil)
  )

(use-package company-box
  :straight t
  :hook (company-mode . company-box-mode))

(use-package company-org-block
  :straight t
  :custom
  (company-org-block-edit-style 'inline)
  :hook ((org-mode . (lambda ()
                       (setq-local company-backends '(company-org-block))
                       (company-mode +1)))))
#+END_SRC
*** consult
#+BEGIN_SRC emacs-lisp
(use-package consult
  :straight t
  :demand t
  ;; :bind (("C-s" . consult-line)
  ;;        ("C-M-l" . consult-imenu)
  ;;        ("C-M-j" . persp-switch-to-buffer*)
  ;;        :map minibuffer-local-map
  ;;        ("C-r" . consult-history))
  :custom
  (completion-in-region-function #'consult-completion-in-region)
  (consult-fontify-preserve nil)
  (consult-async-min-input 0)
  (consult-async-refresh-delay 0.1)
  (consult-async-input-throttle 0.1)
  (consult-async-input-debounce 0.1))
#+END_SRC
*** counsel
#+BEGIN_SRC emacs-lisp
;; Used to prioritize commonly used counsel-M-x commands
(use-package amx
  :straight t)

(use-package counsel
  :straight t
  :ensure-system-package
  ((fzf . fzf)
   (fd . fd)
   (rg . ripgrep))
  :bind (
         :map counsel-describe-map
         ("M-." . counsel-find-symbol)
         :map ivy-minibuffer-map
         )
  :init
  (require 'amx)
  (counsel-mode)

  :config
  ;; (setq counsel-fzf-cmd "rg --files | fzf -f \"%s\"")

  ;; ivy integration
  (with-eval-after-load 'ivy
    (add-to-list 'ivy-more-chars-alist '(counsel-rg . 0))
    (add-to-list 'ivy-more-chars-alist '(counsel-ag . 0))
    ;; remap M-x
    (global-set-key (kbd "M-x") 'counsel-M-x)))
#+END_SRC
*** evil-nerd-commenter
#+BEGIN_SRC emacs-lisp
(use-package evil-nerd-commenter
  :straight t
  :bind ("C-/" . evilnc-comment-or-uncomment-lines))
#+END_SRC
*** doom-themes
#+BEGIN_SRC emacs-lisp
(use-package doom-themes :straight t)
#+END_SRC
*** evil-surround
#+BEGIN_SRC emacs-lisp
(use-package evil-surround
  :straight t
  :after evil
  :config
  (global-evil-surround-mode))
#+END_SRC
*** flx
#+BEGIN_SRC emacs-lisp
(use-package flx
  :straight t
  :after ivy
  :init
  (setq ivy-flx-limit 10000))
#+END_SRC
*** hl-todo
#+BEGIN_SRC emacs-lisp
(use-package hl-todo
  :straight t
  :hook (prog-mode . hl-todo-mode))
#+END_SRC
*** ivy
#+BEGIN_SRC emacs-lisp
(use-package ivy
  :straight t
  :after counsel
  :custom
  (ivy-initial-inputs-alist nil)
  (ivy-extra-directories nil)
  (ivy-wrap t)
  (ivy-count-format "(%d/%d) ")
  ;; (ivy-use-virtual-buffers t)
  (ivy-use-selectable-prompt t)
  (ivy-height 20)
  (ivy-fixed-height-minibuffer t)
  (ivy-re-builders-alist '((t . ivy--regex-fuzzy)))
  :bind (
         :map ivy-minibuffer-map
         ("C-j" . ivy-next-line)
         ("C-k" . ivy-previous-line)
         :map ivy-switch-buffer-map
         ("C-k" . ivy-previous-line)
         ("C-d" . ivy-switch-buffer-kill)
         :map ivy-reverse-i-search-map
         ("C-k" . ivy-previous-line)
         ("C-d" . ivy-reverse-i-search-kill)
         )
  :config
  ;; close with <esc>
  (define-key ivy-minibuffer-map [escape] 'minibuffer-keyboard-quit)
  (ivy-mode 1)
  )
#+END_SRC
*** ivy-rich
#+BEGIN_SRC emacs-lisp
;; (use-package ivy-rich
;;   :straight t
;;   :after ivy
;;   :config
;;   (ivy-rich-mode 1)
;;   (setcdr (assq t ivy-format-functions-alist) #'ivy-format-function-line))
;; (use-package ivy-rich
;;   :straight t
;;   :after counsel
;;   :init
;;   (ivy-rich-mode 1)
;;   :config
;;   (setq ivy-format-function #'ivy-format-function-line)
;;   (setq ivy-rich-display-transformers-list
;;         (plist-put ivy-rich-display-transformers-list
;;                    'ivy-switch-buffer
;;                    '(:columns
;;                      ((ivy-rich-candidate (:width 40))
;;                       (ivy-rich-switch-buffer-indicators (:width 4 :face error :align right)); return the buffer indicators
;;                       (ivy-rich-switch-buffer-major-mode (:width 12 :face warning))          ; return the major mode info
;;                       (ivy-rich-switch-buffer-project (:width 15 :face success))             ; return project name using `projectile'
;;                       (ivy-rich-switch-buffer-path (:width (lambda (x) (ivy-rich-switch-buffer-shorten-path x (ivy-rich-minibuffer-width 0.3))))))  ; return file path relative to project root or `default-directory' if project is nil
;;                      :predicate
;;                      (lambda (cand)
;;                        (if-let ((buffer (get-buffer cand)))
;;                            ;; Don't mess with EXWM buffers
;;                            (with-current-buffer buffer
;;                              (not (derived-mode-p 'exwm-mode)))))))))
(use-package ivy-rich
  :straight t
  :hook ((ivy-mode counsel-mode) . ivy-rich-mode)
  :custom
  (ivy-virtual-abbreviate 'abbreviate)
  (ivy-rich-path-style 'abbrev)
  :config
  (setcdr (assq t ivy-format-functions-alist) #'ivy-format-function-line))
#+END_SRC
*** json-mode
#+BEGIN_SRC emacs-lisp
(use-package json-mode
  :straight t)
#+END_SRC
*** marginalia
#+BEGIN_SRC emacs-lisp
(use-package marginalia
  :after vertico
  :straight t
  :custom
  (marginalia-annotators '(marginalia-annotators-heavy marginalia-annotators-light nil))
  :init
  (marginalia-mode))
#+END_SRC
*** neotree
#+BEGIN_SRC emacs-lisp
(use-package neotree
  :straight t
  :config
  (setq neo-smart-open t)
  ;; (setq projectile-switch-project-action 'neotree-projectile-action)
  (setq neo-theme (if window-system 'icons 'arrows))
  (evil-define-key 'normal neotree-mode-map (kbd "RET") 'neotree-enter)
  (evil-define-key 'normal neotree-mode-map (kbd "TAB") 'neotree-quick-look)
  (evil-define-key 'normal neotree-mode-map (kbd "q") 'neotree-hide)
  (evil-define-key 'normal neotree-mode-map (kbd "C-v") 'neotree-enter-vertical-split)
  (evil-define-key 'normal neotree-mode-map (kbd "C-s") 'neotree-enter-horizontal-split)
  (evil-define-key 'normal neotree-mode-map (kbd "r") 'neotree-refresh)
  (evil-define-key 'normal neotree-mode-map (kbd "h") 'neotree-hidden-file-toggle)
  )

(nmap "-" (lambda ()
            (interactive)
            (if (string-equal major-mode "neotree-mode")
                (select-window (previous-window))
              (neotree-show))
            ))
#+END_SRC
*** org-superstar
#+BEGIN_SRC emacs-lisp
(use-package org-superstar
  :straight t
  :after org
  :hook (org-mode . org-superstar-mode)
  :custom
  (org-superstar-remove-leading-stars t)
  (org-superstar-prettify-item-bullets nil)
  (org-superstar-headline-bullets-list '("◉" "○" "●" "○" "●" "○" "●")))
#+END_SRC
*** paren
Highligth matching braces.
#+BEGIN_SRC emacs-lisp
(use-package paren
  :straight t
  :config
  (set-face-attribute 'show-paren-match-expression nil :background "#363e4a")
  (show-paren-mode 1))
#+END_SRC
*** rainbow-delimiters
#+BEGIN_SRC emacs-lisp
(use-package rainbow-delimiters
  :straight t
  :hook (prog-mode . rainbow-delimiters-mode))
#+END_SRC
*** savehist
Preserve history.
#+BEGIN_SRC emacs-lisp
(use-package savehist
  :straight t
  :init
  (setq history-length t)
  (setq history-delete-duplicates t)
  (savehist-mode t))
#+END_SRC
*** smartparens
#+BEGIN_SRC emacs-lisp
(use-package smartparens
  :straight t
  :hook (prog-mode . smartparens-mode))
#+END_SRC
*** which-key
#+BEGIN_SRC emacs-lisp
(use-package which-key
  :straight t
  :config
  (setq which-key-idle-delay 0.3)
  (setq which-key-popup-type 'frame)
  (which-key-mode)
  (which-key-setup-side-window-bottom)
  (set-face-attribute 'which-key-local-map-description-face nil :weight 'bold))
#+END_SRC
*** undo-tree
#+BEGIN_SRC emacs-lisp
(use-package undo-tree
  :straight t
  :init
  (global-undo-tree-mode 1)
  :custom
  (undo-tree-visualizer-timestamps t)
  (undo-tree-visualizer-diff t)
  :config
  )

(eval-after-load 'undo-tree
  '(progn
     (define-key undo-tree-map (kbd "C-/") nil)
     (define-key undo-tree-map (kbd "C-_") nil)
     (define-key undo-tree-map (kbd "C-?") nil)
     (define-key undo-tree-map (kbd "M-_") nil)
     (define-key undo-tree-map (kbd "C-z") 'undo-tree-undo)
     (define-key undo-tree-map (kbd "C-S-z") 'undo-tree-redo)
     )
  )
#+END_SRC
*** projectile
#+BEGIN_SRC emacs-lisp
(use-package projectile
  :straight t
  :custom
  (projectile-indexing-method 'alien)
  ;; (projectile-enable-caching t)
  (projectile-completion-system 'ivy)
  (projectile-track-known-projects-automatically nil)
  (projectile-sort-order 'recentf)
  (projectile-require-project-root nil)
  (projectile-switch-project-action #'projectile-dired)
  :config
  (projectile-mode t)
  )

(use-package counsel-projectile
  :straight t
  :after projectile
  :config
  (counsel-projectile-mode))
#+END_SRC
*** rainbow-mode
#+BEGIN_SRC emacs-lisp
(use-package rainbow-mode
  :straight t
  :defer t
  :hook (org-mode
         emacs-lisp-mode
         web-mode
         typescript-mode
         js2-mode))
#+END_SRC
* Library
** Development
*** -log
#+BEGIN_SRC emacs-lisp
(defun -log (&rest args)
  (interactive)
  (apply #'message "%S" args)
  )
#+END_SRC
* Visual
** Theme
#+BEGIN_SRC emacs-lisp
(with-eval-after-load 'doom-themes
  (setq
   doom-themes-padded-modeline 3
   doom-themes-enable-italic t
   doom-themes-enable-bold t)
  (doom-themes-neotree-config)
  (doom-themes-org-config)

  (load-theme 'doom-Iosvkem t)
  )
#+END_SRC
** Modeline
Install icons: all-the-icons-install-fonts
#+BEGIN_SRC emacs-lisp
(use-package minions
  :straight t
  :hook (doom-modeline-mode . minions-mode))

(use-package doom-modeline
  :straight t
  :hook (after-init . doom-modeline-mode)
  :custom-face
  (mode-line ((t (:height 100))))
  (mode-line-inactive ((t (:height 100))))
  :custom
  (doom-modeline-height 15)
  (doom-modeline-bar-width 6)
  (doom-modeline-lsp t)
  (doom-modeline-github nil)
  (doom-modeline-mu4e nil)
  (doom-modeline-irc t)
  (doom-modeline-persp-name nil)
  (doom-modeline-buffer-file-name-style 'truncate-except-project)
  (doom-modeline-major-mode-icon nil)
  (doom-modeline-minor-modes nil)
  (doom-modeline-icon t)
  :config
  (line-number-mode 1)
  (column-number-mode 1)
  )
#+END_SRC
** Fonts
#+BEGIN_SRC emacs-lisp
(set-face-attribute 'default nil :font "BMono" :height 100 :weight 'normal)
(set-face-attribute 'fixed-pitch nil :font "BMono" :height 94 :weight 'normal)
(set-face-attribute 'variable-pitch nil :font "Fira Sans" :height 100 :weight 'normal)

(use-package company-posframe
  :config
  (company-posframe-mode 1))

(bind-key "C-+" 'text-scale-increase)
(bind-key "C--" 'text-scale-decrease)
(bind-key "C-0" 'text-scale-adjust)

(setq inhibit-compacting-font-caches t)
#+END_SRC
** Symbol substitution
#+BEGIN_SRC emacs-lisp
(setq-default prettify-symbols-alist '(("#+BEGIN_SRC" . "*")
                                       ("#+END_SRC" . "―")
                                       ("#+begin_src" . "*")
                                       ("#+end_src" . "―")
                                       (">=" . "≥")
                                       ("=>" . "⇨")))
(setq prettify-symbols-unprettify-at-point 'right-edge)

;; Hook org-mode
(add-hook 'org-mode-hook 'prettify-symbols-mode)
#+END_SRC
** Horizontal margins :pkg:
#+BEGIN_SRC emacs-lisp
(use-package olivetti
  :straight t
  :init
  (setq olivetti-body-width .8)
  :config
  (add-hook 'text-mode-hook 'olivetti-mode)
  )
#+END_SRC
** Line numbers
#+BEGIN_SRC emacs-lisp
(dolist (mode '(text-mode-hook prog-mode-hook conf-mode-hook))
  (add-hook mode (lambda () (display-line-numbers-mode 1))))
#+END_SRC
** Highlight trailing white space
#+BEGIN_SRC emacs-lisp
(setq-default show-trailing-whitespace t)

;; exclude
(dolist (hook '(
                special-mode-hook
                term-mode-hook
                comint-mode-hook
                compilation-mode-hook
                minibuffer-setup-hook))
  (add-hook hook
            (lambda () (setq show-trailing-whitespace nil))))
#+END_SRC
** Highlight current line
#+BEGIN_SRC emacs-lisp
(require 'hl-line)
(add-hook 'prog-mode-hook #'hl-line-mode)
(add-hook 'text-mode-hook #'hl-line-mode)
#+END_SRC
* Org
** Setup
*** Org hook
#+BEGIN_SRC emacs-lisp
(defun my/org-hook ()
  (variable-pitch-mode 1)
  (auto-fill-mode 0)
  (visual-line-mode 1)
  (org-indent-mode)
  )
(add-hook 'org-mode-hook #'my/org-hook)
#+END_SRC
*** evil-org
#+BEGIN_SRC emacs-lisp
;; evil-org setup
(use-package evil-org
  :straight t
  :after org
  :hook (org-mode . evil-org-mode)
  :config
  (require 'evil-org-agenda)
  (evil-org-agenda-set-keys))

;; custom (sane) normal mode indent and outdent
(defun my/evil-org-indent ()
  (interactive)
  (evil-org->
   (org-element-property :begin (org-element-at-point))
   (org-element-property :end (org-element-at-point)) 1)
  )
(defun my/evil-org-outdent ()
  (interactive)
  (evil-org->
   (org-element-property :begin (org-element-at-point))
   (org-element-property :end (org-element-at-point)) -1)
  )

;; evil-org hook
(defun my/evil-org-hook ()
  (nmap "d" 'evil-delete)
  (evil-define-key '(normal) 'evil-org-mode
    (kbd ">") 'my/evil-org-indent
    (kbd "<") 'my/evil-org-outdent
    ))
(add-hook 'evil-org-mode-hook #'my/evil-org-hook)
#+END_SRC
*** General
#+BEGIN_SRC emacs-lisp
(setq org-element-use-cache nil) ;; https://www.mail-archive.com/emacs-orgmode@gnu.org/msg140360.html
(setq org-startup-indented t)
(setq org-startup-folded t)
(setq org-catch-invisible-edits 'show-and-error)
(setq org-imenu-depth 999)

(setq org-return-follows-link t)
(setq org-ellipsis " ⤵")
(setq org-hide-emphasis-markers t)
(setq org-fontify-done-headline t)
(setq org-fontify-quote-and-verse-blocks t)
(setq org-pretty-entities t)
(setq org-capture-bookmark nil)
(setq org-outline-path-complete-in-steps nil)
(setq org-refile-use-outline-path t)
(setq org-list-description-max-indent 5)
(setq org-adapt-indentation nil)

(setq-default org-enforce-todo-dependencies t)
(setq-default org-export-with-todo-keywords nil)

;; blocks
(setq org-hide-block-startup nil)
(setq org-src-fontify-natively t)
(setq org-src-preserve-indentation nil)
(setq org-src-window-setup 'current-window)
(setq org-src-tab-acts-natively t)
(setq org-edit-src-content-indentation 2)

;; cycle
(setq org-cycle-separator-lines -1)
;; org-cycle-emulate-tab nil

;; agenda
(setq calendar-week-start-day 1)
#+END_SRC
*** Fixed-pitch faces
#+BEGIN_SRC emacs-lisp
(custom-theme-set-faces
 'user
 '(org-block ((t (:inherit fixed-pitch))))
 '(org-code ((t (:inherit (shadow fixed-pitch)))))
 '(org-document-info-keyword ((t (:inherit (shadow fixed-pitch)))))
 '(org-indent ((t (:inherit (org-hide fixed-pitch)))))
 '(org-meta-line ((t (:inherit (font-lock-comment-face fixed-pitch)))))
 '(org-property-value ((t (:inherit fixed-pitch))) t)
 '(org-special-keyword ((t (:inherit (font-lock-comment-face fixed-pitch)))))
 '(org-table ((t (:inherit fixed-pitch))))
 '(org-tag ((t (:inherit (shadow fixed-pitch) :weight bold :height 0.8))))
 '(org-verbatim ((t (:inherit (shadow fixed-pitch))))))
#+END_SRC
*** Headings
#+BEGIN_SRC emacs-lisp
(set-face-attribute 'org-document-title nil :weight 'bold :height 1.6)
(dolist (face '((org-level-1 . 1.24)
                (org-level-2 . 1.12)
                (org-level-3 . 1.06)
                (org-level-4 . 1.0)
                (org-level-5 . 1.0)
                (org-level-6 . 1.0)
                (org-level-7 . 1.0)
                (org-level-8 . 1.0)))
  (set-face-attribute (car face) nil :weight 'bold :height (cdr face)))

;; remove the background on column views
(set-face-attribute 'org-column nil :background nil)

dotfiles/nvim/lua/config/options.lua
local env = require("lib/env")

return {
  -- base
  clipboard = "unnamed,unnamedplus",
  fileformat = "unix",
  hidden = true,
  lazyredraw = false,
  modeline = false,
  mouse = "nv",
  secure = true,
  shell = vim.fn.exepath("bash"),
  shortmess = "filnxtToOFcs",
  termguicolors = true,
  title = true,
  visualbell = true,
  wrap = false,
  linebreak = true,
  updatetime = 0,
  laststatus = 3,
  scrolloff = 8,
  sidescrolloff = 16,
  timeout = true,
  timeoutlen = 1000,
  ttimeoutlen = 0,
  -- ttimeout = false,

  -- history and persistence
  backup = false,
  backupdir = env.dirs.vim.backup,
  history = 10000,
  swapfile = false,
  undodir = env.dirs.vim.undo,
  viewdir = env.dirs.vim.view,
  undofile = true,
  undolevels = 10000,
  writebackup = false,
  jumpoptions = "view",
  shada = {
    "!",
    "'1000",
    "<100",
    "s100",
    "h",
  },

  -- sessions
  viewoptions = "cursor,folds",
  -- windows
  splitbelow = true,
  splitright = true,
  equalalways = false,
  splitkeep = "screen",
  winborder = "rounded",
  -- text editing
  autoindent = true,
  backspace = [[indent,eol,start]],
  formatoptions = "tcqjn12", -- "cront",
  joinspaces = false,
  textwidth = 999,
  -- completion
  complete = ".",
  completeopt = [[menu,menuone,noselect]],
  -- search
  gdefault = true,
  ignorecase = true,
  inccommand = "split",
  magic = true,
  smartcase = true,
  grepprg = [[rg --glob "!.git" --no-heading --vimgrep --follow $*]],
  -- folds
  -- foldcolumn = "1",
  foldenable = true,
  foldlevel = 999,
  foldlevelstart = 999,
  foldopen = "mark,percent,quickfix,search,tag,undo",
  -- foldmethod = "expr",
  -- foldexpr = "v:lua.vim.treesitter.foldexpr()", -- this fucks up tree-sitter folds
  -- foldexpr = "nvim_treesitter#foldexpr()",
  -- foldnestmax = 1,
  -- code style
  expandtab = true,
  shiftround = true,
  shiftwidth = 2,
  tabstop = 2,
  report = 500,
  -- visual
  showtabline = 1,
  cmdheight = 1,
  concealcursor = "c",
  conceallevel = 2,
  cursorline = true,
  cursorlineopt = "number,screenline",
  fillchars = {
    eob = " ",
    horiz = "━",
    horizup = "┻",
    horizdown = "┳",
    vert = "┃",
    vertleft = "┫",
    vertright = "┣",
    verthoriz = "╋",
    fold = " ",
  },
  list = true,
  ruler = false,
  listchars = {
    space = " ",
    tab = "  ",
    trail = "·",
    nbsp = "␣",
    extends = "›",
    precedes = "‹",
  },
  number = true,
  pumblend = 0,
  pumheight = 12,
  showmode = false,
  signcolumn = "yes",
  synmaxcol = 200,
  smoothscroll = true,
  -- diff
  diffopt = [[hiddenoff,iwhiteall,algorithm:patience]],
}
